File: app/src/main/java/me/raatiniemi/worker/features/project/view/ProjectActivity.java
/*
 * Copyright (C) 2018 Tobias Raatiniemi
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 2 of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package me.raatiniemi.worker.features.project.view;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;

import me.raatiniemi.worker.Preferences;
import me.raatiniemi.worker.R;
import me.raatiniemi.worker.features.project.timesheet.view.TimesheetFragment;
import me.raatiniemi.worker.features.shared.view.activity.BaseActivity;
import me.raatiniemi.worker.util.KeyValueStore;

import static me.raatiniemi.worker.util.NullUtil.isNull;
import static me.raatiniemi.worker.util.NullUtil.nonNull;

public class ProjectActivity extends BaseActivity {
    public static final String MESSAGE_PROJECT_ID = "project id";

    /**
     * Tag for the timesheet fragment.
     */
    private static final String FRAGMENT_TIMESHEET_TAG = "timesheet";

    private final Preferences preferences = new Preferences();
    private final KeyValueStore keyValueStore = preferences.getKeyValueStore();

    /**
     * Reference to the timesheet fragment.
     */
    private TimesheetFragment timesheetFragment;

    public static Intent newIntent(Context context, Long projectId) {
        Intent intent = new Intent(context, ProjectActivity.class);
        intent.setFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP);
        intent.putExtra(ProjectActivity.MESSAGE_PROJECT_ID, projectId);

        return intent;
    }

    /**
     * Get the timesheet fragment, handles construction if needed.
     *
     * @return Timesheet fragment.
     */
    private TimesheetFragment getTimesheetFragment() {
        if (isNull(timesheetFragment)) {
            timesheetFragment = TimesheetFragment.newInstance(getIntent().getExtras());
        }

        return timesheetFragment;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_project);

        if (isNull(savedInstanceState)) {
            getSupportFragmentManager().beginTransaction()
                    .replace(
                            R.id.fragment_container,
                            getTimesheetFragment(),
                            ProjectActivity.FRAGMENT_TIMESHEET_TAG
                    )
                    .commit();
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.actions_project, menu);

        // Set the selected value for the option, otherwise the value will be set to default each
        // time the activity is created.
        MenuItem hideRegistered = menu.findItem(R.id.actions_project_hide_registered);
        if (nonNull(hideRegistered)) {
            hideRegistered.setChecked(keyValueStore.hideRegisteredTime());
        }

        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (R.id.actions_project_hide_registered == item.getItemId()) {
            handleHideRegisteredTimeChange(item);
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    private void handleHideRegisteredTimeChange(MenuItem item) {
        item.setChecked(!item.isChecked());

        keyValueStore.setHideRegisteredTime(item.isChecked());
        getTimesheetFragment().refresh();
    }
}

File: app/src/main/java/me/raatiniemi/worker/features/project/view/ProjectActivity.kt
/*
 * Copyright (C) 2018 Tobias Raatiniemi
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 2 of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package me.raatiniemi.worker.features.project.view

import android.content.Context
import android.content.Intent
import android.os.Bundle
import androidx.lifecycle.Observer
import me.raatiniemi.worker.R
import me.raatiniemi.worker.domain.model.Project
import me.raatiniemi.worker.features.project.model.ProjectHolder
import me.raatiniemi.worker.features.project.timereport.view.TimeReportFragment
import me.raatiniemi.worker.features.shared.view.activity.BaseActivity
import me.raatiniemi.worker.util.NullUtil.isNull
import org.koin.android.ext.android.inject

class ProjectActivity : BaseActivity() {
    private val projectHolder: ProjectHolder by inject()

    private lateinit var timeReportFragment: TimeReportFragment

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_project)

        projectHolder += readProject(intent)
        projectHolder.value.observe(this, Observer {
            title = it.name
        })

        timeReportFragment = TimeReportFragment.newInstance()

        if (isNull(savedInstanceState)) {
            supportFragmentManager.beginTransaction()
                    .replace(
                            R.id.flFragmentContainer,
                            timeReportFragment,
                            ProjectActivity.FRAGMENT_TIME_REPORT_TAG
                    )
                    .commit()
        }
    }

    private fun readProject(intent: Intent): Project {
        return intent.run {
            val id = getLongExtra(MESSAGE_PROJECT_ID, 0)
            val name = getStringExtra(MESSAGE_PROJECT_NAME)

            Project(id, name)
        }
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)

        val newProject = readProject(intent)
        projectHolder.value.run {
            val project = value ?: return@run

            if (project != newProject) {
                projectHolder += newProject
                timeReportFragment.reloadTimeReport()
            }
        }
    }

    companion object {
        private const val MESSAGE_PROJECT_ID = "project id"
        private const val MESSAGE_PROJECT_NAME = "project name"

        /**
         * Tag for the time report fragment.
         */
        private const val FRAGMENT_TIME_REPORT_TAG = "time report"

        @JvmStatic
        fun newIntent(context: Context, project: Project): Intent {
            val intent = Intent(context, ProjectActivity::class.java)
            intent.flags = Intent.FLAG_ACTIVITY_SINGLE_TOP
            intent.putExtra(ProjectActivity.MESSAGE_PROJECT_ID, project.id)
            intent.putExtra(ProjectActivity.MESSAGE_PROJECT_NAME, project.name)

            return intent
        }
    }
}
